### PR TITLE
Adding image support to list view

### LINKED ISSUE #19866: A11y_Azure Data Studio extension 'Azure SQL Migration'_View/Select_Object Information: Issues displayed in assessments results dialog are defined as table instead of list items
"[Check out Accessibility Insights! ](https://accessibilityinsights.io/)- Identify accessibility bugs before check-in and make bug fixing faster and easier.”

### Github Tags:
#A11yMAS; #A11yTCS; #A11ySev2; #Benchmark; #SH-AzureSQLMigration-Win32-Jun2022; #DesktopApp; #Win32; #SQL Migration Control Plane; #WCAG4.2.1; #Object Information;

### Environment and OS details:
Azure Data Studio Version 1.37.0(User Setup)
Windows Version 21H2 (OS Build 22000.675)

### Repro Steps:

1. Turn on Narrator/NVDA
2. Install Azure data studio app and install Azure SQL migration extension
3. Connect to SQL server, tab till respective server, right click and select manage
4. Tab till Azure SQL Migration in sitemap and press enter key
5. Tab till migrate to azure sql and press enter
6. Navigate to databases grid in "Databases for assessment", select few databases and click next.
7. In Assessment results and recommendations step, select a azure sql target and navigate to "View/select" button and press enter
8. After selecting a database in "Assessment results" dialog, issues will be displayed.
9. Verify that issues displayed in assessments results dialog are defined as table

### Actual Result:
Issues displayed in assessments results dialog are defined as table instead of list items

### Expected Result:
Issues displayed in assessments results dialog should be defined as list items instead of giving a empty column headers and empty column 1.

### Attachment:
table_

### PR DESCRIPTION
**Pull request by aasimkhan30**
This PR fixes #19866

Before:
image

After:
image

### CODE REVIEW

File: src/sql/azdata.proposed.d.ts [hunk 1]
```diff
@@ -539,6 +539,21 @@ declare module 'azdata' {
...
+		 */
+		ariaLabel?: string;
+		/**
+		 * Specify the icon for the option. The value could the path to the icon or and ADS icon defined in {@link SqlThemeIcon}.
+		 */
+		icon?: IconPath;
+		/**
+		 * Add icon padding to the list item
+		 */
+		addIconPadding?: boolean;
```
Comment (alanrenmsft) [inline src/sql/azdata.proposed.d.ts:554]: we shouldn't need a separate property for this, if the icon is specified, then the padding should be added automatically.

Comment (aasimkhan30) [reply src/sql/azdata.proposed.d.ts:554]: This is to add forceIconPadding. In some situations, list option may not have icons but we want them to be aligned to ones that have the icons.

Comment (alanrenmsft) [reply src/sql/azdata.proposed.d.ts:554]: if that is the case I would prefer a different name, e.g. addIndention.

Comment (alanrenmsft) [reply src/sql/azdata.proposed.d.ts:554]: or, we can still automatically whether any of the item in the list has icon specified, if so we add padding for all items so that they are aligned.

Comment (Charles-Gagnon) [reply src/sql/azdata.proposed.d.ts:554]: I like that second option - I can't imagine that someone would want to have it not be indented in that case.

File: extensions/sql-migration/src/dialog/assessmentResults/assessmentResultsDialog.ts [hunk 2]
```diff
@@ -45,9 +45,13 @@ export class AssessmentResultsDialog {
 		return new Promise<void>((resolve, reject) => {
 			dialog.registerContent(async (view) => {
 				try {
+					/**
+					 * Subtracting 33px as it removes vertical spacebar caused due 100% height
```
Comment (alanrenmsft) [inline extensions/sql-migration/src/dialog/assessmentResults/assessmentResultsDialog.ts:49]: can you rephrase this?

File: src/sql/workbench/browser/modelComponents/listView.component.ts [hunk 3]
```diff
@@ -18,6 +19,15 @@ import { IListRenderer, IListVirtualDelegate } from 'vs/base/browser/ui/list/lis
...
 import { ScrollbarVisibility } from 'vs/base/common/scrollable';
 import { ILogService } from 'vs/platform/log/common/log';
+import { createIconCssClass } from 'sql/workbench/browser/modelComponents/iconUtils';
+
+interface ListViewOption {
+	id: string;
+	label: string;
+	ariaLabel?: string;
+	icon?: azdata.IconPath;
+	addIconPadding?: boolean;
```
Comment (alanrenmsft) [inline src/sql/workbench/browser/modelComponents/listView.component.ts:29]: this can be removed, right?

Comment (aasimkhan30) [reply src/sql/workbench/browser/modelComponents/listView.component.ts:29]: I still need to add a flag to list item template to add the additional padding.

File: extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts [hunk 4]
```diff
@@ -138,9 +138,9 @@ export class ResourceTypePickerDialog extends DialogBase {
 				text: loc.resourceTypeCategoryListViewTitle
 			},
 			CSSStyles: {
-				'width': '140px',
 				'margin-top': '35px'
 			},
+			width: '140px',
```
[the remark this answers is not in the document]
Comment (aasimkhan30) [reply extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts:143]: The height and width were not getting applied. I have fixed it in the component.

File: src/sql/workbench/browser/modelComponents/listView.component.ts (hunk 3 above)
Comment (alanrenmsft) [reply src/sql/workbench/browser/modelComponents/listView.component.ts:29]: you are using the icon information to decide the value of addIconPadding, which is redundant.

Comment (aasimkhan30) [reply src/sql/workbench/browser/modelComponents/listView.component.ts:29]: If in case only some options have icons, the icons would be undefined for the other options. This will just make sure that the List option renderer is forced to add icon padding.

Comment (alanrenmsft) [reply src/sql/workbench/browser/modelComponents/listView.component.ts:29]: my point is: it is a calculated field, you shouldn't need to expose it in the interface:
image

Comment (aasimkhan30) [reply src/sql/workbench/browser/modelComponents/listView.component.ts:29]: This is not exposed to the extensions and is only local to the component. I have removed this property from azdata.proposed.ts.

Comment (alanrenmsft) [reply src/sql/workbench/browser/modelComponents/listView.component.ts:29]: yes, I know, but the places that might use this component still see this property and will be confused, you should have an internal flag somewhere to track whether icon padding should be added for the list.

Comment (alanrenmsft) [reply src/sql/workbench/browser/modelComponents/listView.component.ts:29]: if it is not very straightforward to implement, please make this one optional as well as add comment explaining the why it is here and can't be removed.

File: src/sql/workbench/browser/modelComponents/listView.component.ts [hunk 5]
```diff
@@ -193,11 +193,21 @@ interface ExtensionListTemplate {
 
 class OptionsListRenderer implements IListRenderer<ListViewOption, ExtensionListTemplate> {
 	public static TEMPLATE_ID = 'optionListRenderer';
+	private _forceIconPadding: boolean = false;
 
 	public get templateId(): string {
 		return OptionsListRenderer.TEMPLATE_ID;
 	}
 
+	public set forceIconPadding(v: boolean) {
```
Comment (alanrenmsft) [inline src/sql/workbench/browser/modelComponents/listView.component.ts:195]: looks like all you need is a public property: `public forceIconPadding : boolean = false;`

**Review by alanrenmsft — APPROVED** [review]

File: src/sql/workbench/browser/modelComponents/listView.component.ts [hunk 6]
```diff
@@ -80,11 +87,16 @@ export default class ListViewComponent extends ComponentBase<azdata.ListViewComp
...
 	override ngOnDestroy(): void {
+		this.iconClasses.forEach((icon) => {
+			DOM.removeCSSRulesContainingSelector(icon);
+		});
 		this.baseDestroy();
 	}
 
 	public get options(): azdata.ListViewOption[] {
-		return this.getProperties().options ?? [];
+		const options = this.getProperties().options ?? [];
```
Comment (Charles-Gagnon) [inline src/sql/workbench/browser/modelComponents/listView.component.ts:97]: 1. It's odd to only set forceIconPadding when getting options since it shouldn't change unless the properties change. So I'd suggest adding something like this to a setProperties override instead
2. Instead of `forceIconPadding` I'd suggest something more generic like `displayIcon` (and then in the doc comment make clear this will show not only icons but also empty spaces for options without icons)
3. And even better IMO, you could make a CSS rule for showing/hiding the icon fields and then just set the CSS class here on the component itself instead of having the renderer do more calculation

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
See comment for some suggestions on making this cleaner

**Comment by aasimkhan30** [comment]
@Charles-Gagnon, updated code to use your approach. Thanks for the help!

File: src/sql/workbench/browser/modelComponents/listView.component.ts [hunk 7]
```diff
@@ -28,10 +30,13 @@ export default class ListViewComponent extends ComponentBase<azdata.ListViewComp
 	@Input() modelStore: IModelStore;
 	@ViewChild('vscodelist', { read: ElementRef }) private _vscodeList: ElementRef;
 	private _optionsList!: List<azdata.ListViewOption>;
+	private _optionsListRenderer: OptionsListRenderer;
 	private _selectedElementIdx!: number;
 
 	static ROW_HEIGHT = 26;
 
+	private iconClasses: string[] = [];
```
Comment (Charles-Gagnon) [inline src/sql/workbench/browser/modelComponents/listView.component.ts:38]: It looks like this isn't used anymore

**Review by Charles-Gagnon — APPROVED** [review]
